### PR TITLE
Low battery

### LINKED ISSUE #49: Implement Low Battery Flight Mode
Low Battery mode currently raises NotImplementedError. Time to implement it!
See https://cornell.app.box.com/file/655915541214 for a (brief) description.

### PR DESCRIPTION
**Pull request by tmf97**
Closes #49 
Gives a rough draft of our low battery flight mode

### CODE REVIEW

File: telemetry/telemetry.py [hunk 1]
```diff
@@ -78,6 +81,18 @@ def get_acc(self):
     def get_mag(self):
         return self.mag
 
+    def write(self):
+        pass
```
Comment (sjz38) [inline telemetry/telemetry.py:85]: Is this a #TODO?

**Review by sjz38 — CHANGES_REQUESTED** [review]

File: flight_modes/flight_mode.py [hunk 2]
```diff
@@ -351,30 +344,98 @@ def __init__(self, parent):
...
+        if time_for_opnav:  # TODO: and FMEnum.OpNav.value not in self._parent.FMQueue:
+            self._parent.FMQueue.put(FMEnum.OpNav.value)
+
+        if time_for_telem:
+            telem = self._parent.telemetry.minimal_packet()
+            downlink = self._parent.downlink_handler.pack_downlink(self._parent.downlink_counter,
+                                                                   FMEnum.LowBatterySafety.value,
+                                                                   LowBatterySafetyCommandEnum.CritTelem.value,
+                                                                   **telem)
+            self._parent.downlink_queue.put(downlink)
```
[the remark this answers is not in the document]
Comment (sjz38) [reply flight_modes/flight_mode.py:429]: Is it that we can still run certain tasks in low battery mode? I guess we do (like maneuver, etc), but are we in the hopes that the battery won't drain to 0 for these tasks?

[the remark this answers is not in the document]
Comment (tmf97) [reply flight_modes/flight_mode.py:429]: We want to still be able to do tasks in low batt mode, just at a lower frequency to preserve power.

File: flight_modes/flight_mode.py [hunk 3]
```diff
@@ -351,30 +344,98 @@ def __init__(self, parent):
         super().__init__(parent)
         raise NotImplementedError
 
-    def update_state(self):
+    def update_state(self) -> int:
         return NO_FM_CHANGE  # intentional: we don't want to update FM when testing sensors
 
 
 class LowBatterySafetyMode(FlightMode):
+    """FMID 3: Low Battery Safety Mode: the main goal here is to save power as much as possible in case we are in a
```
[the remark this answers is not in the document]
Comment (tmf97) [reply flight_modes/flight_mode.py:352]: On it!

File: telemetry/telemetry.py (hunk 1 above)
Comment (tmf97) [reply telemetry/telemetry.py:85]: yup

File: flight_modes/flight_mode.py [hunk 4]
```diff
@@ -351,30 +344,98 @@ def __init__(self, parent):
...
 
     def run_mode(self):
-        sleep(params.LOW_BATT_MODE_SLEEP)  # saves battery, maybe?
-        raise NotImplementedError
+        if self.task_completed:
+            sleep(params.LOW_BATT_MODE_SLEEP)  # saves battery, maybe?
+        else:
+            if self._parent.gom is not None:
+                self._parent.gom.all_off()  # turns everything off initially upon entering mode to preserve power
+                self._parent.gom.rf_transmitting_switch(
```
[the remark this answers is not in the document]
Comment (tmf97) [reply flight_modes/flight_mode.py:399]: Ah yeah good point

File: flight_modes/flight_mode.py [hunk 5]
```diff
@@ -642,10 +712,20 @@ def update_state(self):
             logger.info("Time to run Opnav")
             return FMEnum.OpNav.value
 
+        if time_for_telem:
+            telem = self._parent.telemetry.standard_packet_dict()
+            downlink = self._parent.downlink_handler.pack_downlink(
+                self._parent.downlink_counter, FMEnum.Normal.value, NormalCommandEnum.BasicTelem.value,
+                **telem)
+
+            self._parent.downlink_queue.put(downlink)
```
Comment (sjz38) [inline flight_modes/flight_mode.py:721]: Should we update `self._parent.last_telem_downlink` here as well? Just looking at what is done in low battery mode

File: telemetry/telemetry.py [hunk 6]
```diff
@@ -327,3 +350,8 @@ def write_telem(self):
             self.session.commit()
         finally:
             self.session.close()
+
+    def query_telem(self):
```
Comment (sjz38) [inline telemetry/telemetry.py:354]: Add # TODO

**Review by sjz38 — CHANGES_REQUESTED** [review]
See comments for the questions I have

File: main.py [hunk 7]
```diff
@@ -216,7 +217,7 @@ def poll_inputs(self):
         # TODO: move this following if block to the telemetry file
         if self.radio is not None:
             # Telemetry downlink
-            if (datetime.today() - self.radio.last_telemetry_time).total_seconds() / 60 >= params.TELEM_DOWNLINK_TIME:
+            if (datetime.today() - self.radio.last_telemetry_time).total_seconds() / 60 >= params.TELEM_INTERVAL:
```
[the remark this answers is not in the document]
Comment (tmf97) [reply main.py:220]: Oh huh, yeah they should be time() values. Also, we add to the downlink queue both in the main loop and in the flight modes - I think we should only do so in the flight modes.

File: flight_modes/low_battery.py [hunk 8]
```diff
@@ -9,20 +10,25 @@ class LowBatterySafetyMode(FlightMode):
...
     downlink_codecs = {
-        LowBatterySafetyCommandEnum.BasicTelem.value: ([RTC_TIME, ATT_1, ATT_2, ATT_3, ATT_4,
-                                                        HK_TEMP_1, HK_TEMP_2, HK_TEMP_3, HK_TEMP_4, GYRO_TEMP,
-                                                        THERMOCOUPLER_TEMP,
-                                                        CURRENT_IN_1, CURRENT_IN_2, CURRENT_IN_3,
-                                                        VBOOST_1, VBOOST_2, VBOOST_3, SYSTEM_CURRENT, BATTERY_VOLTAGE,
-                                                        PROP_TANK_PRESSURE], 84),
-
-        LowBatterySafetyCommandEnum.CritTelem.value: (
+        NormalCommandEnum.BasicTelem.value: ([RTC_TIME, ATT_1, ATT_2, ATT_3, ATT_4,
```
Comment (sjz38) [inline flight_modes/low_battery.py:24]: Why are we switching this to `NormalCommandEnum`?

File: telemetry/telemetry.py [hunk 9]
```diff
@@ -283,6 +304,80 @@ def minimal_packet(self):
                 BATT_MODE: self.gom.hk.battmode,
                 PPT_MODE: self.gom.hk.pptmode}
 
+    def detailed_packet_dict(self) -> Dict[str, Union[int, float]]:
```
Comment (sjz38) [inline telemetry/telemetry.py:307]: Seems like some changes are becoming out of the scope of the Low battery branch. If there are adjustments that need to be made to telemetry, then we should set up a new branch and add it to Jira

**Review by sjz38 — APPROVED** [review]
Looks good

File: flight_modes/low_battery.py (hunk 8 above)
Comment (tmf97) [reply flight_modes/low_battery.py]: Ah I was doing a hacky workaround on code that I didn't fully know how it worked (even though I wrote it) to pass the tests

File: telemetry/telemetry.py (hunk 9 above)
Comment (tmf97) [reply telemetry/telemetry.py:307]: Kinda, but not too far outta scope. For low battery we need a smaller packet to downlink, and this is just that implementation. But yes I'm very guilty of scope creep

**Pull request merged**